### ISSUE TITLE
Feature Request - Queuing Request Logs

### ISSUE THREAD

**Issue by Sammyjo20** [bilfeldt/laravel-request-logger#30]
Hey, absolutely love this package - great work on it!

We're currently running into an issue where because every request is written to the `request_logs` table. This results in a single point of failure (if our writer instance fails). It also means that for even read-based operations our writer is only being used. It would be really cool if this library allowed us to queue the storing of request logs. This would allow our GET requests to only touch our load-balanced reader instances and never touch the writer. It would also speed up our requests as we currently have the request logs table on a separate database connection.

**Comment by bilfeldt** [comment]
Hi @Sammyjo20 👋 

> absolutely love this package - great work on it!

Thanks so much 🙏 

> We're currently running into an issue where because every request is written to the request_logs table. This results in a single point of failure (if our writer instance fails). It also means that for even read-based operations our writer is only being used. It would be really cool if this library allowed us to queue the storing of request logs. This would allow our GET requests to only touch our load-balanced reader instances and never touch the writer. It would also speed up our requests as we currently have the request logs table on a separate database connection.

I see your use case 🤔  I originally avoided queued jobs and looked at [Terminable Middleware](https://laravel.com/docs/10.x/middleware#terminable-middleware) but finally decided to use a synchronous listener (aka non-queued) for the event `Illuminate\Foundation\Http\Events\RequestHandled`.

## Solution right now

Could we not just queue the listener we already have for the `Illuminate\Foundation\Http\Events\RequestHandled`? No - [because the `Request` object is not serializable](https://github.com/laravel/framework/issues/11551#issuecomment-167404376).

That means that we need to **call** the logger (implementing [`RequestLoggerInterface`](https://github.com/bilfeldt/laravel-request-logger/blob/main/src/Contracts/RequestLoggerInterface.php)) synchronously, but there **is no problem in using a logger that does the actual writing using a queued job**.

This can actually be done already, in two ways (option 2 is easier):
1. Add your own driver which uses a queued job. Add this driver to the [config](https://github.com/bilfeldt/laravel-request-logger/blob/2fe2f6ebc48b607f83c8c05de9926b8ed68369d1/config/request-logger.php#L57) and set that as the [default](https://github.com/bilfeldt/laravel-request-logger/blob/2fe2f6ebc48b607f83c8c05de9926b8ed68369d1/config/request-logger.php#L45)
2. Make a new model which extends `\Bilfeldt\RequestLogger\Models\RequestLog` and simply override the log function to queue the actual saving. Then [configure that the package should use your new child class instead](https://github.com/bilfeldt/laravel-request-logger/blob/2fe2f6ebc48b607f83c8c05de9926b8ed68369d1/config/request-logger.php#L59C24-L59C72).
 

## Should we implement it into this package as a core feature?

Albeit it is already possible, then there is no build in config for it, which would be the ideal for any developer wanting to achieve this.

I originally thought there were no use cases for this, since the performance concern makes no sense simply because writing is done **after** the response has been returned to the user. But I do understand your concerns and could imagine there could be other edgecases.

**Hence I am okay adding this as a build in feature of the package.** I will keep this issue open for now, and if I find the time to add this myself I will probably do so. But I am open for any PRs for this feature already.

**Comment by Sammyjo20** [comment]
Hey @bilfeldt I'm really sorry for not replying sooner - I feel awful!

Thank you for such a detailed reply. It would be great if this was a feature in the library but I understand you may have other commitments at this point in time. I wanted to share this with you - we wanted to see what was slowing some of our requests down and it was what I expected:

![Screenshot 2024-02-27 at 12 28 15](https://github.com/bilfeldt/laravel-request-logger/assets/29132017/05488aca-e644-429a-ac03-31607b84d5ec)

**Comment by bilfeldt** [comment]
No worries at all @Sammyjo20 :)

Just to be clear, that query is being executed *after* the response has been returned to the user right?

**Comment by Sammyjo20** [comment]
No, the middleware isn't terminatable - so it'll be part of the request - this is why it's being picked up in the request lifecycle

**Comment by bilfeldt** [comment]
> No, the middleware isn't terminatable - so it'll be part of the request - this is why it's being picked up in the request lifecycle

The middleware is only used to "tag" that the request should be logged. Logging happens using the event listener here https://github.com/bilfeldt/laravel-request-logger/blob/2fe2f6ebc48b607f83c8c05de9926b8ed68369d1/src/EventServiceProvider.php#L11-L15

And the event `RequestHandled` is fired **after** the response has been returned - right?

**Comment by Sammyjo20** [comment]
Hmm, the listener isn't queued so that leaves me to think it might be running synchronously? I might be wrong - where abouts is the `RequestHandled` event being fired?

**Comment by Sammyjo20** [comment]
Sorry - I didn't realise it was a Laravel event: 

https://github.com/laravel/framework/blob/988206bcacf79829bda66af4a016778e4b4d762a/src/Illuminate/Foundation/Http/Kernel.php#L152

But it doesn't look like this is after the request lifecycle. I'll keep digging haha

**Comment by bilfeldt** [comment]
Yes the listener is not queued (and cannot be, because of the serialization issue I mentioned).

I might have misunderstood the `RequestHandled` event. I was almost certain it fired **after** the response was returned to the user, but your reference would indicate I am wrong there 😒 

If that is indeed the case then we/I should **definitly** make it possible to log sync either using a queued job or using terminatable middleware.

**Comment by bilfeldt** [comment]
Sorry @Sammyjo20 had not seen your answer when I posted above.

So we agree that my original postulate was **wrong** and that the event `RequestHandled` is fired **before** the response is returned to the user correct?

I wonder if there is a `RequestTerminated` event we could listen for 🤔

**Comment by Sammyjo20** [comment]
I'm currently working off a fork and I will see if it works - just installing it locally - the PR will be a little crude without testing but I am happy to put one up if you like? Thank you so much for being so cooperative on this one!

**Comment by bilfeldt** [comment]
> I'm currently working off a fork and I will see if it works - just installing it locally - the PR will be a little crude without testing but I am happy to put one up if you like? Thank you so much for being so cooperative on this one!

Feel free to post a PR for this - absolutely 👍 

A few ideas I have here:

1. Simply add a new `queued` config and then simply move this part into a queued job: https://github.com/bilfeldt/laravel-request-logger/blob/2fe2f6ebc48b607f83c8c05de9926b8ed68369d1/src/Models/RequestLog.php#L97-L124
2. Instead of listening for the `RequestHandled` event, then we register a `terminatedCallback`: https://github.com/laravel/framework/blob/7c680bb58dfb15923209045f99ed5cf46c31f45d/src/Illuminate/Foundation/Application.php#L1354-L1364
3. Instead of using the `RequestHandled` event then we use a terminatable middleware - that would be a breaking change, because it would require everybody to use that middleware, and there are currently other ways to enable logging.

**Comment by bilfeldt** [comment]
I personally vote for combining option 1 and 2 btw 👍

**Comment by Sammyjo20** [comment]
I think either option 1 or 2 could be good. For people who haven't enabled queueing it will still run in sync, but for others it'll queue it. Option 2 is probably easier to implement because instead of an event being fired, you could run the terminatable middleware which checks for a request. 

I'd be happy to try and change my PR to option 2 if you would like.

**Comment by Sammyjo20** [comment]
With option 1 you could even `Job::dispatchAfterResponse()` which would possibly reduce latency even further.

**Comment by bilfeldt** [comment]
Good idea. Yes a PR would be much appriciated.

**Comment by michaelchora** [comment]
@Sammyjo20 @bilfeldt - this was a really interesting read. We are also facing a similar issue with the huge number of requests the app gets per second. 

did this feature ever get a PR? We are keen to implement queueing for our logging too.

**Comment by bilfeldt** [comment]
This has not been PR'ed - but feel free 👍 

After re-evaluating this, I have come to dislike queueing because pushing logging to the queue means pushing both the `Request` and the `Response` objects to a queued job. But both of these objects can contain sensitive data like auth or personal data which can then be exposed in the queue driver (often a database).

Instead using **[Deferred functions](https://laravel.com/docs/12.x/helpers#deferred-functions)** is my preferred solution for now as it provides the performance benefits without any performance downsides.

> [!NOTE]
> It is already possible to push the logging to **after** the response has been sent to the uses using deferred functions. Simple   implement your own Eloquent Model (extending `Bilfeldt\RequestLogger\Models\RequestLog`) and override the `log` function to defer the call to the parent:
>```php
>class RequestLog extends Bilfeldt\RequestLogger\Models\RequestLog
>{
>    /** @inerhitDoc */
>    public function log(Request $request, $response, ?int $duration = null, ?int $memory = null): void
>    {
>        defer(fn () => parent::log($request, $response, $duration, $memory))->always();
>     }
>}
>```

**Comment by onlime** [comment]
> Instead using **[Deferred functions](https://laravel.com/docs/12.x/helpers#deferred-functions)** is my preferred solution for now as it provides the performance benefits without any performance downsides.

agree.

> It is already possible to push the logging to **after** the response has been sent to the uses using deferred functions. Simple implement your own Eloquent Model (extending `Bilfeldt\RequestLogger\Models\RequestLog`) and override the `log` function to defer the call to the parent:

@bilfeldt Any reason why you didn't want to make this the default in your package? maybe too obtrusive?
I decided to make deferred logging the default in [`onlime/laravel-sql-reporter` v1.3.2](https://github.com/onlime/laravel-sql-reporter/releases/tag/v1.3.2)

**Comment by bilfeldt** [comment]
> @bilfeldt Any reason why you didn't want to make this the default in your package? maybe too obtrusive?

Nope, I think it should be, but I would like it to be configurable.

Not sure if it should be considered a breaking change though.
